deals/deals.go

```diff
@@ -225,6 +234,195 @@ func (m *Module) Watch(ctx context.Context, proposals []cid.Cid) (<-chan DealInf
+		return
+	}
+	for _, dr := range pendingDeals {
+		remaining := time.Until(time.Unix(dr.Time, 0).Add(dealTimeout))
+		if remaining <= 0 {
+			go m.finalizePendingDeal(dr)
+		} else {
+			go m.eventuallyFinalizeDeal(dr, remaining)
+		}
+	}
+}
+
+func (m *Module) recordDeal(params *api.StartDealParams, proposalCid cid.Cid) {
+	di := DealInfo{
+		PieceCID:      params.Data.Root,
+		Duration:      params.MinBlocksDuration,
+		PricePerEpoch: params.EpochPrice.Uint64(),
+		Miner:         params.Miner.String(),
+		ProposalCid:   proposalCid,
+	}
+	record := DealRecord{
+		Addr:     params.Wallet.String(),
+		Time:     time.Now().Unix(),
+		DealInfo: di,
+		Pending:  true,
+	}
+	log.Infof("storing pending deal record for proposal cid: %s", proposalCid.String())
+	if err := m.store.putPendingDeal(record); err != nil {
+		log.Errorf("storing pending deal: %v", err)
+		return
+	}
+	go m.eventuallyFinalizeDeal(record, dealTimeout)
+}
+
+func (m *Module) finalizePendingDeal(dr DealRecord) {
+	deletePending := func() {
+		if err := m.store.deletePendingDeal(dr.DealInfo.ProposalCid); err != nil {
+			log.Errorf("deleting pending deal for proposal cid %s: %v", dr.DealInfo.ProposalCid.String(), err)
+		}
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	info, err := m.api.ClientGetDealInfo(ctx, dr.DealInfo.ProposalCid)
+	if err != nil {
+		log.Errorf("getting deal info: %v", err)
+		deletePending()
+		return
+	}
+	if info.State != storagemarket.StorageDealActive {
+		log.Infof("pending deal for proposal cid %s isn't active yet, deleting pending deal", dr.DealInfo.ProposalCid.String())
+		deletePending()
+	} else {
+		di, err := fromLotusDealInfo(ctx, m.api, info)
+		if err != nil {
+			log.Errorf("converting proposal cid %s from lotus deal info: %v", dr.DealInfo.ProposalCid.String(), err)
+			deletePending()
+			return
+		}
+		record := DealRecord{
+			Addr:     dr.Addr,
+			Time:     time.Now().Unix(), // Note: This can be much later in time than the deal actually became active on chain
+			DealInfo: di,
+			Pending:  false,
+		}
+		if err := m.store.putFinalDeal(record); err != nil {
+			log.Errorf("storing propoosal cid %s deal record: %v", dr.DealInfo.ProposalCid.String(), err)
+		}
+	}
+}
+
+func (m *Module) eventuallyFinalizeDeal(dr DealRecord, timeout time.Duration) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	updates, err := m.Watch(ctx, []cid.Cid{dr.DealInfo.ProposalCid})
+	if err != nil {
+		log.Errorf("watching proposal cid %s: %v", dr.DealInfo.ProposalCid.String(), err)
+		return
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			log.Infof("watching proposal cid %s timed out, deleting pending deal", dr.DealInfo.ProposalCid.String())
+			if err := m.store.deletePendingDeal(dr.DealInfo.ProposalCid); err != nil {
+				log.Errorf("deleting pending deal: %v", err)
+			}
+			return
+		case info, ok := <-updates:
+			if !ok {
```

Comment:
Not sure if checking `ok` is overkill here since the channel only exists within the local scope, but did it anyway